### ISSUE TITLE
dataset's map operation performance is O(N) with data shape for O(1) operations, can be orders of magnitude slower than using a dataset from generator.

### ISSUE THREAD

**Issue by feature-engineer** [tensorflow/tensorflow#35135]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux 18.04, Windows 7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0
- Python version: 3.7.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
dataset.map takes 10 times longer when getting 10 times bigger tensors in the dataset, regardless of the function being passed. e.g. `data.map(lambda x: x[0, 0])` would take 100 times longer, if `data` contained tensors of shape `(1000, 1000)` rather than ` (100, 100)`.

**Describe the expected behavior**
I would expect that if the mapped function has an O(1) performance, mapping it on any dataset would still have O(1) performance not O(N).

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
timeit.timeit("""
tf.reduce_sum([x for x in data.map(lambda x: x[:10, :10]).repeat().take(1000)])
""",
setup="""
import tensorflow as tf
tf.autograph.set_verbosity(0, False)
x = tf.ones((1, 10000,1000))
data = tf.data.Dataset.from_tensor_slices([x])
""", number=10)
```
takes 10 times longer than 

```
timeit.timeit("""
tf.reduce_sum([x for x in data.map(lambda x: x[:10, :10]).repeat().take(1000)])
""",
setup="""
import tensorflow as tf
tf.autograph.set_verbosity(0, False)
x = tf.ones((1, 1000,1000))
data = tf.data.Dataset.from_tensor_slices([x])
""", number=10)
```

Where the only difference is x having 10 times the values in the first example.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by feature-engineer** [comment]
For comparison, 
```
timeit.timeit("""
tf.reduce_sum([x for x in data.take(1000)])
""",
setup="""
import tensorflow as tf
def generator(x):
  def gen():
    yield x[:10, :10]
  return gen
x = tf.ones((1000,10000), dtype=tf.float32)
data = tf.data.Dataset.from_generator(generator(x), tf.float32)
""", number=10)
```
runs 1000 times faster than the equivalent using map, as written in my initial example.

**Comment by amahendrakar** [comment]
Able to reproduce the issue. Please find the [Gist](https://colab.sandbox.google.com/gist/amahendrakar/16937cdaede037dd39886d1cf07f0f9d/35135.ipynb) here. Thanks!

**Comment by jsimsa** [comment]
Your benchmark is not measuring the speed of `map` execution accurately. `from_tensor_slices` performs a data copy so executing your input pipeline with inputs of different size will result in different amount of data being copied (which would on its own explain the O(1) vs O(n) difference).

A better way to test whether `map` is O(1) (which it should be for constant-time operations) vs. O(n) is to use `tf.data.Dataset.from_tensors(X).repeat()` varying the size of `X` in your setup.

Now as for the `from_generator` comparison, Python-based `from_generator` is expected to perform better than tf.data `map` (which incurs about 10us overhead per element to execute the user-defined computation through TensorFlow executor). The benefits of using `map` are realized when you are performing non-trivial computation that can benefit from either the graph optimizations TensorFlow performs and / or parallel execution enabled through the `num_parallel_calls` argument.

**Comment by feature-engineer** [comment]
@jsimsa `from_tensor_slices ` is performed in the setup, and isn't counted in the timing loop.. adding .repeat() to it on setup doesn't change the result.

This is a toy example, to make it clear that the problem is with the .map operation, I've started off with a training of a simple network comparing different sized tensors which are randomly cropped. The performance in training is proportional to the tensor size, which should not be the case. See [issue #34936 ](https://github.com/tensorflow/tensorflow/issues/34936).

As for comparing to from_generator, I understand there's a small overhead to using map, but this overhead shouldn't be dependant on the input sizes, and in my example, it is.

**Comment by jsimsa** [comment]
Your setup only builds the input pipeline graph. The `from_tensor_slices` data copy is performed in each call to `get_next` which only happens in your `tf.reduce_sum`. The following [presentation](https://www.youtube.com/watch?v=kVEOCfBy9uY) explains in detail how tf.data input pipeline construction and execution works.

Your `gist` is still using `from_tensor_slices`. Can you reproduce the issue by using `from_tensors` with differently sized inputs?

**Comment by feature-engineer** [comment]
Ok, from_tensors does indeed perform with O(1), while from_tensor_slices has an O(N) performance.

Is this the intended behaviour? If so, it should be mentioned in the documentation, as this is a very significant difference between these otherwise quite similar methods.

**Comment by jsimsa** [comment]
I am generally opposed to documenting implementation details (e.g. which methods are implementated using a data copy and which are not).

@aaudiber what are your thoughts on this? This is related to our discussion of deprecating `from_tensor_slices()` in favor of `from_tensors().unbatch()` -- if we do that, should we document that `unbatch()` performs a data copy?

**Comment by aaudiber** [comment]
I am also generally opposed to exposing implementation details since we don't want people to rely on the specific implementation and then get in trouble when the implementation changes.

In cases where the performance is worse than users might expect, I think it's worth including a note along the lines of 

Note: `unbatch()` performs a copy of the input data, so it can be expensive. Try to avoid unnecessary calls to `unbatch()`.

If we update the implementation to avoid the copy in the future, we can update the doc and (hopefully) no one will complain that we've saved a copy.

**Comment by jsimsa** [comment]
So do you think we should update the documentation of all tf.data transformations that currently perform a data copy (`batch`, `padded_batch`, `from_tensor_slices`, `unbatch`, ...)?

**Comment by aaudiber** [comment]
I don't think calling out the data copy is important for `batch`, `padded_batch`, or `from_tensor_slices` because those copies usually aren't the root cause of performance issues.  But we have had issues where liberal use of `unbatch` causes performance issues, so we should call out the performance considerations.

**Comment by feature-engineer** [comment]
Well, if you've got big images you want to semantically segment, you could be tempted to use `from_tensor_slices` and map to random_crop. 

That was my scenario. I wasn't aware that it'd be so inefficient when compared to `from_tensors`, since it wasn't mentioned anywhere. I searched the slowdown in other places until I was told here that it's `from_tensor_slices` that is to blame. 

I don't see why anyone would want to use `from_tensor_slices` when it copies the input tensors on each iteration, or why it's the version more commonly used in the examples and guides.

In my opinion, either get rid of `from_tensor_slices` completely, or add a note regarding the aberrant performance - correct me if I'm wrong, but no other method of dataset creation entails a copy of its input on each iteration.

**Comment by jsimsa** [comment]
@aaudiber `from_tensor_slices` is not that different from `from_tensors` followed by `unbatch`. For datasets materialized into memory (e.g. as a NumPy array), `from_tensor_slices` is commonly used in TF tutorials as an example to convert these into a dataset and this will be suboptimal (c.f. this issue).

@feature-engineer `from_tensor_slices` represents a "ease-of-use vs. performance" trade-off and TF tutorials optimize for simplicity over performance. In many cases, the overhead of the data copy is negligible compared to the per-element pre-processing computation. For best performance, not only you want to avoid / minimize data copies and vectorize and parallelize processing. An example of a performant and complex use of tf.data for handling NumPy objects can be found in [keras](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/data_adapter.py#L215) which converts NumPy objects to a tf.data.Dataset that shuffles and batches the original data (in a way that outperforms doing the same in NumPy).

As a side note, is there a reason you are materializing your images in memory outside of tf.data as opposed to reading the images (presumably from files) using tf.data ([example](https://www.tensorflow.org/guide/data#consuming_sets_of_files))?

**Comment by feature-engineer** [comment]
@jsimsa I get my training data from a stream over the network, so it's already loaded to memory.. If it's too big to include in the graph, I create a temporary TFRecord out of it, and build a dataset from that. But if the training data is small enough, I skip the TFRecord creation and create the dataset straight from the in-memory data. It's a semantic segmentation model, so the images are big, and I randomly crop them during training (after randomly selecting the validation data and masking it out).

**Comment by jsimsa** [comment]
Thank you for the explanation. In the case the data is small and present in memory, you can use `from_generator` to avoid the `from_tensor_slices` data copy.

I am going to close this issue as the original reason for opening (O(n) time complexity of `map`) has been refuted and the O(n) time complexity of `from_tensor_slices` is expected. 

@aaudiber as per our offline discussion, it makes sense to comment on the `from_tensor_slices` and `unbatch` performance in either the API documentation of the tf.data [performance guide](https://www.tensorflow.org/guide/data_performance).

@feature-engineer feel free to create a PR updating the [performance guide](https://github.com/tensorflow/docs/blob/master/site/en/guide/data_performance.ipynb) with information that would help others with a similar use case. Thank you.
